godot-core/src/builtin/arrays.rs

```diff
@@ -494,16 +577,46 @@ impl fmt::Debug for Array {
     }
 }
 
+/// Creates a new reference to the data in this array. Changes to the original array will be
+/// reflected in the clone and vice versa.
+///
+/// To create a (mostly) independent copy instead, see [`Array::duplicate_shallow()`] and
+/// [`Array::duplicate_deep()`].
+impl Clone for Array {
+    fn clone(&self) -> Self {
+        unsafe {
+            Self::from_sys_init(|self_ptr| {
+                let ctor = ::godot_ffi::builtin_fn!(array_construct_copy);
+                let args = [self.sys_const()];
+                ctor(self_ptr, args.as_ptr());
+            })
+        }
+    }
+}
+
```

Comment:
This is the default implementation that is provided by the `impl_builtin_traits`, which you removed: 🙂 
```rs
( Clone for $Type:ty => $gd_method:ident ) => {
    impl Clone for $Type {
        #[inline]
        fn clone(&self) -> Self {
            unsafe {
                Self::from_sys_init(|self_ptr| {
                    let ctor = ::godot_ffi::builtin_fn!($gd_method);
                    let args = [self.sys_const()];
                    ctor(self_ptr, args.as_ptr());
                })
            }
        }
    }
};
```

But didn't we want to omit `Clone` until the semantics are clarified and consistent across all the Godot data structures (`Gd`, `Dictionary`, `Array`, `Packed*Array`, `String`, `StringName`)?

Reply:
It's the same... except for the documentation.

Might be nice to have `impl_builtin_trait` (singular) which also accepts metadata (i.e. docstrings).

It's important to have this functionality (whatever it's called), because API methods take `Array` by value.

My takeaway from the discussion was that there's no way to prevent `Array`'s reference semantics from leaking into our API. Also, `String` and `StringName` already have `Clone`-by-reference semantics (albeit undocumented), so this is consistent.

How about this: I DRY up this impl, I add a section to the docs of the `builtin` module about what's going on, and I link to that from `Array` docs?

Reply:
Are people looking up trait `impl`s in the documentation? In 95% of cases, you only see the default documentation from `std`. I think this would be better mentioned in the struct's documentation, then it's also harder to miss it.

> It's important to have this functionality (whatever it's called), because API methods take `Array` by value.

They can use one of the `duplicate_*` methods for that until then.

> Also, `String` and `StringName` already have `Clone`-by-reference semantics (albeit undocumented), so this is consistent.

The difference for strings is that they're immutable, so this behavior is 100% transparent to the user. There are no traps like "modifications of the old string will silently sneak into copies", at least I'm not aware of them 😬 

> How about this: I DRY up this impl, I add a section to the docs of the `builtin` module about what's going on, and I link to that from Array docs?

What I want to avoid is that we solidify a certain meaning of `Clone` now, only to break it in a couple of weeks. Or worse, we leave an inconsistent API like https://github.com/godot-rust/gdnative/issues/712, where people need to look up the meaning for each type. If people are taught "`Clone` is always cheap" and then we suddenly change it to perform deep copies, that will cause nasty surprises. This topic clearly needs more discussion and I'd prefer not to block this PR because of it, as the `duplicate_*` methods are enough to unblock users for now.

What you could consider for now is use the `obj::traits::Share` trait in its place -- this should also be easy to change later, if we conclude that it's the best fit for `Clone`. This would be consistent with `Gd<T>`.

Reply:
Implemented `Share`!